### ISSUE TITLE
type 'HttpException' is not a subtype of type 'FileInfo'

### ISSUE THREAD

**Issue by Moncader** [renefloor/flutter_cache_manager#99]
In the newest release (1.1.0),
https://github.com/renefloor/flutter_cache_manager/blob/master/lib/src/cache_manager.dart#L120
Here you are yielding the exception. However in the case when the download failed an HttpException is what is being thrown, and the Stream doesn't can't cast that exception to a FileInfo.

This breaks any kind of network error for this plugin.

**Comment by renefloor** [comment]
Could you give me an example of the issue? 
When I use the example in `cached_network_image` it works as expected. Also without an internet connection.

**Comment by Moncader** [comment]
The reason you aren't seeing it is because you aren't looking at the exception produced and are ignoring it (we log our exceptions for serious errors outside of HttpException)

https://github.com/renefloor/flutter_cached_network_image/blob/master/lib/src/cached_image_widget.dart#L275

If you put a breakpoint there and look at the value of the `snapshot.error` you'll notice it isn't an HttpException, it's a `_TypeError` due to trying to yield an object that isn't the object the stream is expecting.

https://dart.dev/tutorials/language/streams#error-events

**Comment by renefloor** [comment]
Of course! You are right. I'm working on a better solution, expect to make it this week.

**Comment by renefloor** [comment]
@Moncader what do you think of this solution? Idea is to not just return the error, but return it within a FileInfo object: https://github.com/renefloor/flutter_cache_manager/commit/855f306104066db93812ff7217f6b14c455625ae

Application in cached_network_image: https://github.com/renefloor/flutter_cached_network_image/commit/00f1fa940fd1df65e47ecc634e84c5aab1509173

**Comment by larssn** [comment]
Glad we're not the only ones who ran into this error. It seemed to appear out of nowhere. Thanks for trying to fix it.

**Comment by Moncader** [comment]
@renefloor 
Sorry. I haven't had time to prepare a sample, but isn't the 'official' dart way to do it to `throw` the exception, and in your stream listener properly do a try catch? Right now it's all inlined because you're doing a `.where()` after the `read`. I think if you just wrap that with a try catch it would fine?

**Comment by renefloor** [comment]
I used to throw the error, but some users complained that it made there app crash. I don't really know how or why. I'd expect the error to be handled by the stream widget. Normally that works fine. Maybe the error is thrown when the view is already disposed. When it is disposed it shouldn't return anything anymore and especially not throw the error.

Do you know if we could listen to the stream being disposed when we created it like this?

**Comment by larssn** [comment]
It shouldn't crash anyone's app. Inside an `async*` function, throwing errors is the correct way to do it. 

Maybe those users are using VSCode, which can intercept errors thrown, which might make it look like a real crash.

**Comment by Moncader** [comment]
It's not about VSCode. Most major applications need to catch the errors generated by the program and use things like zone error catching and overriding the flutter global error handler to catch things.

Inside here the those errors also get caught (We were getting HTTP 404 errors bubbling up there when we don't care about 404 since there was no way to catch them.).

You should be able to revert back to 'throw' and just properly catch the error wherever you are handling the stream.

I'm pretty sure the problem is here https://github.com/renefloor/flutter_cached_network_image/blob/master/lib/src/cached_image_widget.dart#L271

The Stream is being filtered by a where condition and the so before the stream's data is given to flutter, it's dead from the exception thrown because you are 'where'ing on an exception. (At least I think that's what's going on)

You'll probably need to filter inside the streambuilder itself.

**Comment by alextekartik** [comment]
I encoutered a similar error (so the source of the issue is a network issue):

```
12:08:42.988 3313 info flutter.tools I/flutter (25495): 	CERTIFICATE_VERIFY_FAILED: certificate has expired(handshake.cc:352))
12:08:43.555 3316 info flutter.tools E/flutter (25495): [ERROR:flutter/lib/ui/ui_dart_state.cc(148)] Unhandled Exception: type 'HandshakeException' is not a subtype of type 'FileInfo'
12:08:43.555 3317 info flutter.tools E/flutter (25495): #0      BaseCacheManager.getFile (package:flutter_cache_manager/src/cache_manager.dart:120:11)
```

Indeed it is yielding an error instead of an object of FileInfo here: https://github.com/renefloor/flutter_cache_manager/blob/94b2ff970bdd5f7d30735d1711e65c8a8f714dee/lib/src/cache_manager.dart#L118-L122

I cannot tell what the proper behavior should be...

**Comment by renefloor** [comment]
> You should be able to revert back to 'throw' and just properly catch the error wherever you are handling the stream.
> 
> I'm pretty sure the problem is here https://github.com/renefloor/flutter_cached_network_image/blob/master/lib/src/cached_image_widget.dart#L271
> 
> The Stream is being filtered by a where condition and the so before the stream's data is given to flutter, it's dead from the exception thrown because you are 'where'ing on an exception. (At least I think that's what's going on)
> 
> You'll probably need to filter inside the streambuilder itself.

I had gotten the same idea. Will revert this quickly and see how I can fix the stream in cached_network_image.

**Comment by renefloor** [comment]
Released fix as 1.1.1
